### ISSUE TITLE
Extremely slow while building xformers from source

### ISSUE THREAD

**Issue by jiapei100** [facebookresearch/xformers#822]
# ❓ Questions and Help
Building **xformers** from source extremely slows down my computer ... freezing almost everything.
How much memory does **xformers** need to be built from source?

**Comment by Tostino** [comment]
I had to create extra swap space on my machine to get xformers or flash_attn to build from source and not crash entirely after multiple hours of being unresponsive.

With an extra 64gb swap partition + 96gb of ram, I am able to build, but it takes ~5-10 min to build on my 7950X

**Comment by danthe3rd** [comment]
Hi,
This is a known issue, and it can be mitigated by limiting the number of parallel tasks for build.
Just set `MAX_JOBS=4` before building for instance

**Comment by paboum** [comment]
> This is a known issue, and it can be mitigated by limiting the number of parallel tasks for build.

One can also prevent fork bombs by limiting user processes to a relatively low number or running suspicious code inside virtual machines. But regular people don't just run fork bombs from time to time, I expect default `MAX_JOBS` to be set to 4 by default and if you begin to use a build system that respects the amount of free resources, then you can think of increasing it.

**Comment by youkaichao** [comment]
I find I must set `MAX_JOBS=4` to prevent the building process to crash my server. Then it requires more than one hour to build :(

I'm using a server with 2 L4 GPUs and 24 CPUs with 96GB memory.

**Comment by youkaichao** [comment]
I hope in the future there can be some futures like building only a subset of ops 🥺

**Comment by paboum** [comment]
> I find I must set `MAX_JOBS=4` to prevent the building process to crash my server. 

As I said, this build is configured to act like a fork bomb by default. Half a year has passed and the most innovative team in one of the top valued companies cannot even respond to that. Hopefully more lawsuits emerge.

> Then it requires more than one hour to build :(
> I'm using a server with 2 L4 GPUs and 24 CPUs with 96GB memory.

The above being said, it doesn't take that long on my regular modern PC, maybe 10 mins or so. Perhaps you get lots of warnings underneath, maybe you should read them. Try running setup via `python setup.py install` with some verbosity flags? IDK, I'm not the support here.

**Comment by youkaichao** [comment]
> it doesn't take that long on my regular modern PC, maybe 10 mins or so. 

You do? From my side, I can see it seems to do something like kernel parameter tuning, with `ptax` searching for register allocation plans.

**Comment by alex391** [comment]
Yeah, I'm running into this “the build is a fork bomb” problem as well. Working around it by just setting `MAX_JOBS` to 1, but it would be nice to see this issue get some attention.

**Comment by danthe3rd** [comment]
We're working to make the compilation way faster in the future, but this might take some weeks/months and require a new PT release.
In the meantime, you should be able to use pre-compiled wheels, why are you compiling from source otherwise?

**Comment by alex391** [comment]
I'm trying to get xformers into the package manager [Spack](https://spack.io/), which prefers to build from source.

**Comment by mikeshi80** [comment]
> We're working to make the compilation way faster in the future, but this might take some weeks/months and require a new PT release. In the meantime, you should be able to use pre-compiled wheels, why are you compiling from source otherwise?

For me, I need run it on Nvidia Orin AGX, and there is no pre-compiled wheels for ARM + CUDA platform.
